### PR TITLE
Sidebar Item Drop Inside & External State Handling for Expand

### PR DESCRIPTION
**Pull request by Augmentador**

### CODE REVIEW

**Comment by Augmentador** [comment]
@jmatth Wanted to note that i decided to go with the route of using generic types here. It has increase the quality of our implementation on the composer side of things and it only represented a small change on the example project here on the macos_ui side of things (we just had to type the SidebarItems here with a String type). Happy to go over this during the eng synch if you believe it not to be ideal after the review of both PRs.

**Comment by Augmentador** [comment]
@jmatth Ready for review

File: lib/src/layout/sidebar/sidebar_item.dart [hunk 1]
```diff
@@ -7,7 +8,7 @@ import 'package:macos_ui/src/library.dart';
 ///
 ///  * [Sidebar], a side bar used alongside [MacosScaffold]
 ///  * [SidebarItems], the widget that displays [SidebarItem]s vertically
-class SidebarItem with Diagnosticable {
+class SidebarItem<T extends Object> with Diagnosticable {
```
Comment (jmatth) [inline lib/src/layout/sidebar/sidebar_item.dart:11]: You can just write `<T>`. Every non-null type in Dart extends Object.

File: lib/src/layout/sidebar/sidebar_items.dart [hunk 2]
```diff
@@ -210,19 +210,20 @@ class _SidebarItem extends StatefulWidget {
...
   /// disclousure items of a _DisclosureSidebarItem.
   final bool? isLastDisclousureItem;
 
   @override
-  State<_SidebarItem> createState() => _SidebarItemState();
+  State<_SidebarItem<T>> createState() => _SidebarItemState<T>();
 }
 
-class _SidebarItemState extends State<_SidebarItem> {
+class _SidebarItemState<T extends Object> extends State<_SidebarItem<T>> {
```
Comment (jmatth) [inline lib/src/layout/sidebar/sidebar_items.dart:224]: Same here

File: lib/src/layout/sidebar/sidebar_items.dart [hunk 3]
```diff
@@ -290,7 +292,13 @@ class _SidebarItemState extends State<_SidebarItem> {
         break;
     }
 
-    Widget result = Semantics(
+    Color backgroundColor() {
+      if (_isDraggingInside) return MacosColors.controlAccentColor;
```
Comment (jmatth) [inline lib/src/layout/sidebar/sidebar_items.dart]: Let's use `MacosTheme.of(context).primaryColor` for the drop background and line colors.

**Review by jmatth — COMMENTED** [review]
Just a couple nitpicks, then this should be good to go.

File: lib/src/layout/sidebar/sidebar_item.dart (hunk 1 above)
Comment (Augmentador) [reply lib/src/layout/sidebar/sidebar_item.dart:11]: I did try doing that initially but it results in an issue with the generic type of the DragTarget which does use `T extends Object`

Screenshot 2023-03-14 at 16 04 14

Screenshot 2023-03-14 at 16 04 47

Comment (jmatth) [reply lib/src/layout/sidebar/sidebar_item.dart:11]: Well that's annoying. Nevermind then.

**Pull request merged**